src/cs-utils/FrameStats.cpp

```diff
@@ -0,0 +1,464 @@
+  }
+
+  // Reset the new current pool.
+  auto const& pool = mQueryPools.at(mCurrentQueryPool);
+  pool->reset();
+
+  // Start the "root" full frame timing. This is always done, even if pEnableMeasurements is set to
+  // false. This is required to get data for the pFrameTime property.
+  mFullFrameTimingID = pool->startTimerQuery("Process Frame", FrameStats::TimerMode::eBoth);
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void FrameStats::endFrame() {
+
+  // End the "root" full frame timing. This is always done, even if pEnableMeasurements is set to
+  // false. This is required to get data for the pFrameTime property.
+  mQueryPools.at(mCurrentQueryPool)->endTimerQuery(mFullFrameTimingID);
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+int32_t FrameStats::startTimerQuery(std::string name, FrameStats::TimerMode mode) {
+
+  // Only attempt to start the timing if pEnableMeasurements is set to true.
+  if (pEnableMeasurements.get()) {
+    return mQueryPools.at(mCurrentQueryPool)->startTimerQuery(std::move(name), mode);
+  }
+
+  return -1;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+int32_t FrameStats::startSamplesQuery(std::string name) {
+
+  // Only attempt to start the counting if pEnableMeasurements is set to true.
+  if (pEnableMeasurements.get()) {
+    return mQueryPools.at(mCurrentQueryPool)->startSamplesQuery(std::move(name));
+  }
+
+  return -1;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+int32_t FrameStats::startPrimitivesQuery(std::string name) {
+
+  // Only attempt to start the counting if pEnableMeasurements is set to true.
+  if (pEnableMeasurements.get()) {
+    return mQueryPools.at(mCurrentQueryPool)->startPrimitivesQuery(std::move(name));
+  }
+
+  return -1;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void FrameStats::endTimerQuery(int32_t id) {
+
+  // Only attempt to end the timing if pEnableMeasurements is set to true.
+  if (pEnableMeasurements.get()) {
+    mQueryPools.at(mCurrentQueryPool)->endTimerQuery(id);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void FrameStats::endSamplesQuery(int32_t id) {
+
+  // Only attempt to end the counting if pEnableMeasurements is set to true.
+  if (pEnableMeasurements.get()) {
+    mQueryPools.at(mCurrentQueryPool)->endSamplesQuery(id);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void FrameStats::endPrimitivesQuery(int32_t id) {
+
+  // Only attempt to end the counting if pEnableMeasurements is set to true.
+  if (pEnableMeasurements.get()) {
+    mQueryPools.at(mCurrentQueryPool)->endPrimitivesQuery(id);
+  }
+}
```

Comment:
Does the case exist, where:
1. A query is started
2. The measurements are being disabled
3. The query isn't ended

or does the control flow not allow this scenario?